### PR TITLE
Add store, RPC client, connect store and keymgr to bmclient

### PR DESCRIPTION
**Pull request by ishbir**
Added mailboxes, message storage, pow queue, getpubkey request storage. Connected keymgr and store to bmclient so that database/key file creation and loading works as expected. RPC client for connections to bmd was added. keymgr.LookupByAddress was also added to facilitate retrieval of private identities by address.

Also implemented parsing of PyBitmessage's keys.dat file, added code for storing and loading counter values in/from the data store, decryption of messages and broadcasts, responding to getpubkey requests (untested). Added helper function for generating getpubkey if requested public doesn't identity exist (untested), config option for choosing POW doer and periodic saving of data to keyfile and store.

Package store is a persistent database for everything that bmclient needs for operation. This includes storing mailboxes and messages, proof of work queue, list of requested pubkeys and subscribed broadcast addresses.

Take note that this store does not encrypt everything. Only messages contents are encrypted. Encryption scheme used is SalsaX20 stream cipher with Poly1305 MAC, based on secretbox in NaCl.

WARNING: If both your database and password were compromised, changing your password won't accomplish anything. This is because store encrypts the master key using the password you specify when the database is created. A malicious user that knew your password and had a previous copy of the store knows the master key and can decrypt all messages.

WARNING 2: Master key is kept in memory. The package is susceptible to memory reading malware.

### CODE REVIEW

**Comment by DanielKrawisz** [comment]
Can you make it so that the mailbox stores a unique id with every message that is ever stored in it, as is required by imap?

**Comment by DanielKrawisz** [comment]
How is a store.Message different from a wire.Message? Can you think of another word besides 'message' for it? There's a really terrible ambiguity about this word.

**Comment by DanielKrawisz** [comment]
Also, can you think of a word other than Mailbox? Bitmessage CAN be used like email but it could be used for all kinds of things. I think we should reserve words like 'mailbox' for when we are specifically using it like email.

**Comment by ishbir** [comment]
1. I haven't implemented Mailbox/Message yet. It's just in there so that we can think of an appropriate message format and then implement it.
2. store.Message is an e-mail message that's stored in the 'store'. wire.Message is a message that's sent over the over. I'm not sure what you mean by ambiguity here since Golang takes care of that for us.
3. store.Mailbox is intended to be used ONLY as a 'conventional' e-mail mailbox. If someone wants to use Bitmessage for something other than messages/broadcasts, they're going to have to implement their own store, store whatever they might need, hook it up to bmd and then be able to send all kinds of things out on the network. bmclient is intended to be used only for messaging-like applications of Bitmessage.

**Comment by DanielKrawisz** [comment]
I think we need to take a break and have a discussion because I think we're already doing some redundant work and it would be bad if we had to choose later from completed work. Here are some things I'm thinking about, based on your comments. 

* I have assumed that the information about a message which is stored in bmclient is something compatible with the bitmessage protocol rather than with email formats. In other words, when my email client interacts with bmclient via imap, the message is immediately converted into a bitmessage compatible format. This means that all incompatible headers are stripped. Then, if I were to download the message again, it is reconstructed from a bitmessage-compatible format into an email when it is transferred. If this is not what you had intended, let's pause to discuss it before proceeding. 
* Right now the bitmessage protocol is mainly used in a way that is analogous to email, but it could be expanded easily to be used for other purposes in the future. In particular, other encoding formats might be defined which allow for purposes that won't work like email at all. It is worthwhile to keep this conceptual distinction in mind because part of the goal of bmd is to expand the use of bitmessage generally. I think it is important maintain a conceptual distinction between a) anything compatible with msgmsg or msgbroadcast and b) a msgmsg or msgbroadcast that can be treated like an email. If you disagree then we need to discuss that too. I wrote about this stuff in the big email I sent around last week and I was assuming that since nobody wrote back and disagreed that everyone was on board with it. 
* I have created two interfaces already that may conflict with some of the things you intend to do here but haven't implemented yet. I've named the objects Bitmessage and BitmessageBox (but am open to changing the names.) Bitmessage represents a message that is in a format that can be treated as an email, and BitmessageBox represents a message box that is easily compatible with IMAP.  https://github.com/DanielKrawisz/bmclient/edit/email/message/box.go?pr=/monetas/bmclient/pull/2

I think we may have not separated the work we intended to do clearly enough, so let's fix that right away!

**Comment by ishbir** [comment]
I think you're right. We need to have this discussion as soon as possible. I was too fixated on getting it to work with IMAP and SMTP that I forgot about the larger implications of my intended design (not being able to work with other encoding types). But I think we're still good, I haven't done any work on mailbox and message storage. So the overlap of our work is still very minimal.

**Comment by ishbir** [comment]
As per our discussion, the message store takes in messages as bytes and there is a 'suffix' (which could be message encoding or any other parameter used to look up addresses with that prefix; call it an index, if you will).

**Comment by DanielKrawisz** [comment]
I would like the mailbox to support the following. 
* Get last message in the mailbox with a given suffix. 
* Get all messages with a given suffix between two index numbers. 
* Get all messages with a given suffix >= a given index number. 
Could you add functions for that? 

Also, IMAP requires messages to be stored with a unique id number that never changes even if messages are deleted or reordered. Can we have that in addition to a suffix? If you can add that in, I will also need some functions using that as well. 
* Get last uid assigned. 
* Get message with the highest uid for a given suffix.
* Get message by uid. 
* Get messages with a given suffix in a range of uids.
* Get messages with a given suffix with uid >= a given uid. 
Is that possible? Thanks!

**Comment by ishbir** [comment]
- Change index to uid. What is now called index acts like imap uid, so let's call it uid.
-- I don't see why this cosmetic change should matter but I changed it to ID (to shorten it).

- Function to get last uid assigned.
-- Done

- Function to return the message with the highest uid of a given suffix.
-- Done

- Function to return all messages of a given suffix in a range between two uids inclusive.
  Function to return all messages of a given suffix with uid >= a given value.
  Function to return a slice of all uids assigned of a given suffix.
-- All three tasks can now be done using ForEachMessage.

File: dbsetup.go [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+	valid := []string{"n", "no", "y", "yes"}
+	response, err := promptConsoleList(prefix, valid, defaultEntry)
+	if err != nil {
+		return false, err
+	}
+	return response == "yes" || response == "y", nil
+}
+
+// promptConsolePass prompts the user for a passphrase with the given prefix.
+// The function will ask the user to confirm the passphrase and will repeat
```
Comment (DanielKrawisz) [inline dbsetup.go]: Why is there a prefix that goes with the passphrase?

Comment (ishbir) [inline dbsetup.go]: Prefix means prompt. So, "Enter password for store" is a prefix.

Comment (DanielKrawisz) [inline dbsetup.go]: Oh ok. Why wouldn't you just call it prompt?

Comment (ishbir) [inline dbsetup.go]: I just copied this code over from btcwallet. Didn't bother with renaming anything.

File: store/mailbox.go [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+//
+// Suffix is useful for getting all messages of a particular type. For example,
+// retrieving all messages with encoding type 2.
+//
+// The function terminates early if an error occurs and iterates in the
+// increasing order of index. Make sure it doesn't take long to execute. DO NOT
+// execute any other database operations in it.
+func (mbox *Mailbox) ForEachMessage(lowID, highID, suffix uint64,
+	f func(id, suffix uint64, msg []byte) error) error {
+	if lowID > highID {
```
Comment (DanielKrawisz) [inline store/mailbox.go]: If ForEachMessage only iterates over messages with a given suffix, then it shouldn't be necessary for f to include suffix as an argument, right?

File: store/db.go [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+		// different name.
+		return mbox, err
+	}
+
+	return mbox, nil
+}
+
+// MailboxByAddress retrieves the mailbox associated with an address. If the
+// mailbox doesn't exist, an error is returned.
+func (s *Store) MailboxByAddress(address string) (*Mailbox, error) {
```
Comment (DanielKrawisz) [inline store/db.go]: Does address mean bitmessage address?

File: store/mailbox.go (hunk 2 above)
Comment (ishbir) [inline store/mailbox.go]: I'm not sure what you mean. You could specify suffix as 0 and iterate over all messages.

File: store/db.go (hunk 3 above)
Comment (ishbir) [inline store/db.go]: Yes.

File: store/mailbox.go (hunk 2 above)
Comment (DanielKrawisz) [inline store/mailbox.go]: Oh right. I missed that.

**Pull request merged**